text/0000-float-semantics.md

```diff
@@ -0,0 +1,272 @@
+(See above for why the RFC does not propose a fully deterministic spec for floating-point operations.)
+
+To summarize, achieving deterministic floating-point behavior is too hard with the current state of the art; the downsides of accepting non-determinism are low; so it is not worth blocking floating-point operations in `const fn` on this issue.
+
+Note that this RFC does *not* imply a reliable way for code to detect whether it runs at compile-time or run-time.
+
+### Alternative: `const` just fails when a NaN would be produced
+
+Another alternative to handling floating-point operations in `const` is to just fail when a non-deterministic choice would occur at runtime, i.e., each time a NaN is produced.
+However, this is a breaking change: `const C: f32 = 0.0 / 0.0;` has worked on stable Rust since Rust 1.0.
+
+Even under this alternative we would allow `const fn` to perform operations that are non-deterministic at run-time, i.e., unsafe code could still not rely on `const fn` as being a marker for "this is deterministic at run-time". Thus it shares all the downsides with the previous alternative, it just avoids making NaN choices observable in `const`.
+
+The core advantage of this option is that it avoids having `const` results change when the unspecified compile-time NaN changes on a compiler update or across compilers.
+However, having `const` results depend on NaN bits should be very rare, and we already have other (more common) cases of `const` results depending on unspecified implementation details that can and sometimes do change on compiler updates, namely the layout of `repr(Rust)` types (observable via `size_of` and `offset_of`).
+
+### Alternative: `const` tracks NaN values, fails when their bits matter during compile-time
+
+`const` could in principle track NaNs symbolically, similar to how it tracks pointers, and delay choosing NaN payload bits until codegen.
+Const-evaluation would then abort if the bits of a NaN are observed (eg. if `to_bits`is called during const-evaluation).
+This would keep `const C = 0.0/0.0;` working, but requires `const fn is_nan` to be an intrinsic.
+However, it would require massive amounts of work in the compile-time interpreter, comparable in complexity to all the work that is already required to support symbolic pointers (and the RFC author doubts that there will be a lot of opportunity for those two kinds of symbolic state to share infrastructure).
+That effort should only be invested if there is a significant payoff.
+The RFC author considers the downsides of unspecified NaN bit patterns being observable in const-evaluation to be minimal, and hence the payoff of this alternative to be low.
+
+# Prior art
+[prior-art]: #prior-art
+
+C and C++ generally avoid saying anything about signaling NaNs, and also don't talk about how NaNs are chosen. I can only assume this means they are chosen non-deterministically; that is certainly the interpretation chosen by GCC and LLVM.
```

Comment:
> C and C++ generally avoid saying anything about signaling NaNs

This isn't quite accurate. C17 and prior has explicit wording

> This specification does not define the behavior of signaling NaNs

which makes it undefined behavior (although this tends to be in the sense of "extension point for implementers" rather than "assume it can't happen"). C23 changes this wording to say that

> Where specification of signaling NaNs is not provided, the behavior of signaling NaNs is implementation-defined (either treated as an IEC 60559 quiet NaN or treated as an IEC 60559 signaling NaN).

LLVM presently is closer to the former definition.

Reply:
> This specification does not define the behavior of signaling NaNs

That's what I meant when I said they avoid saying things about it. It's not clear to me that this really means that creating a signaling NaN via bit reinterpretation is UB.

I didn't realize C23 is updating that, thanks for pointing that out! I will update the RFC.

Reply:
(update done)